src/main/java/knightminer/simplytea/item/TeapotItem.java

```diff
@@ -4,69 +4,132 @@
+import net.minecraftforge.common.capabilities.ForgeCapabilities;
+import net.minecraftforge.fluids.FluidActionResult;
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.fluids.FluidType;
+import net.minecraftforge.fluids.FluidUtil;
+import net.minecraftforge.fluids.capability.IFluidHandler;
+
+import java.util.Optional;
 
 public class TeapotItem extends TooltipItem {
+
 	public TeapotItem(Properties props) {
 		super(props);
 	}
 
+	// using onItemUseFirst() instead of use() because many fluid tanks will consume the interaction before it gets to use()
 	@Override
-	public InteractionResultHolder<ItemStack> use(Level world, Player player, InteractionHand hand) {
-		ItemStack stack = player.getItemInHand(hand);
+	public InteractionResult onItemUseFirst(ItemStack stack, UseOnContext context)
+    {
+		Level world = context.getLevel();
+		if (world.isClientSide) {
+        	return InteractionResult.PASS;
+		}
+
+		Player player = context.getPlayer();
+		// Even though the context already has a blockhitresult, it does not include fluid blocks
 		BlockHitResult rayTrace = getPlayerPOVHitResult(world, player, ClipContext.Fluid.SOURCE_ONLY);
 		if (rayTrace.getType() == Type.BLOCK) {
 			BlockPos pos = rayTrace.getBlockPos();
 			BlockState state = world.getBlockState(pos);
+			Block block = state.getBlock();
+			Direction side = rayTrace.getDirection();
+
+			// unable to modify the block
+			if (!world.mayInteract(player, pos) || !player.mayUseItemAt(pos.relative(side), side, stack)) {
+				return InteractionResult.PASS;
+			}
+
+			// cauldron disabled in config
+			if (block instanceof AbstractCauldronBlock && !Config.SERVER.teapot.fillFromCauldron()) {
+				return InteractionResult.PASS;
+			}
+
+			Optional<Item> filledItem = Optional.empty();
+			Optional<SoundEvent> pickupSound = Optional.empty();
+			FluidState fluidState = state.getFluidState();
+
+			if (fluidState.isSource()) {
+				Fluid fluid = fluidState.getType();
+				// comparing with WATER fluid instead of WATER fluid tag, because many/most modded fluids are tagged as WATER
+				// to get water-like interactions even if the fluid is not water
+				if (fluid.isSame(Fluids.WATER)) {
+					filledItem = Optional.of(Registration.teapot_water);
+					pickupSound = ((BucketPickup)block).getPickupSound(state);
+					if (!Config.SERVER.teapot.infiniteWater()) {
+						((BucketPickup)block).pickupBlock(world, pos, state);
+					}
+				} else if (fluid.is(Tags.Fluids.MILK)) {
+					filledItem = Optional.of(Registration.teapot_milk);
+					pickupSound = ((BucketPickup)block).getPickupSound(state);
+					((BucketPickup)block).pickupBlock(world, pos, state);
+				}
+			} else if (Config.SERVER.teapot.fillFromFluidTank()) {
+				BlockEntity entity = world.getBlockEntity(pos);
+				Optional<IFluidHandler> fluidHandler = Optional.empty();
+				if (entity != null) {
+					 fluidHandler = entity.getCapability(ForgeCapabilities.FLUID_HANDLER, side).resolve();
```

Comment:
Don't check the block fluid handler; just make the teapot item expose a fluid handler. See vanilla water buckets as an example.

We won't be able to use the same fluid handler exactly, but should be able to use something similar. See [this class from Tinkers' Construct](https://github.com/SlimeKnights/TinkersConstruct/blob/1.19.2/src/main/java/slimeknights/tconstruct/fluids/util/ConstantFluidContainerWrapper.java#L17) for reference, [here is how to use it](https://github.com/SlimeKnights/TinkersConstruct/blob/1.19.2/src/main/java/slimeknights/tconstruct/fluids/item/ContainerFoodItem.java#L77-L89). For our sake we don't need all that supplier stuff, could just statically create the fluid stack as long as it copies in the right spots.

Reply:
I'd argue that sounds like a bug with the mod that adds MilkJarBlock, it would likely also break other mods of mine like Ceramics. We can probably leave that case as not working for now unless its easy to fix that without hardcoding/adding redundant logic.

Reply:
It's probably due to milk not being a real fluid in Vanilla. Maybe I'll look into submitting a PR for Cooking for Blockheads to see if it can be improved from that side.